File: module_3/java/LongestConsecutive.java
/*
 *   Title: Longest consecutive sequence
 *
 *   Problem:Given an unsorted array of integers, find the length of the
 *   longest consecutive elements sequence.
 *
 *   Your algorithm should run in O(n) complexity.
 *
 *   Execution: javac LongestConsecutive.java && java LongestConsecutive
 */
import java.util.*;


public class LongestConsecutive {
    private static boolean arrayContains(int[] arr, int num) {
        for (int i = 0; i < arr.length; i++) {
            if (arr[i] == num) {
                return true;
            }
        }
        return false;
    }

    public static int longestConsecutiveBF(int[] nums) {
        /* Brute-force implementation of longest consecutive sequence. */
        int longestStreak = 0;

        for (int num : nums) {
            int currentNum = num;
            int currentStreak = 1;

            while (arrayContains(nums, currentNum + 1)) {
                currentNum += 1;
                currentStreak += 1;
            }

            longestStreak = Math.max(longestStreak, currentStreak);
        }
        return longestStreak;
    }

    public static int longestConsecutiveSorting(int[] nums) {
        /* Sorting implementation of longest consecutive sequence. */
        if (nums.length == 0) {
            return 0;
        }

        Arrays.sort(nums);

        int longestStreak = 1;
        int currentStreak = 1;

        for (int i = 1; i < nums.length; i++) {
            if (nums[i] != nums[i-1]) {
                if (nums[i] == nums[i-1]+1) {
                    currentStreak += 1;
                }
                else {
                    longestStreak = Math.max(longestStreak, currentStreak);
                    currentStreak = 1;
                }
            }
        }
        return Math.max(longestStreak, currentStreak);
    }

    public static int longestConsecutiveHashSet(int[] nums) {
        /* Hash-set implementation of longest consecutive sequence. */
        Set<Integer> num_set = new HashSet<Integer>();
        for (int num : nums) {
            num_set.add(num);
        }

        int longestStreak = 0;

        for (int num : num_set) {
            if (!num_set.contains(num-1)) {
                int currentNum = num;
                int currentStreak = 1;

                while (num_set.contains(currentNum+1)) {
                    currentNum += 1;
                    currentStreak += 1;
                }

                longestStreak = Math.max(longestStreak, currentStreak);
            }
        }
        return longestStreak;
    }

    public static void main(String[] args) {
        int[] testArr = {100, 4, 200, 1, 3, 2};
        assert longestConsecutiveBF(testArr) == 4;
        assert longestConsecutiveSorting(testArr) == 4;
        assert longestConsecutiveHashSet(testArr) == 4;
        System.out.println("Explanation: The longest consecutive elements sequence is [1, 2, 3, 4]. Therefore its length is 4.");

        System.out.println("Passed all test cases");
    }
    
}

File: module_3/python/longest_consecutive.py
"""
Title: Longest consecutive sequence.

Problem: 
    Given an unsorted array of integers, find the length of the longest
    consecutive elements sequence.

    Your algorithm should run in O(n) complexity.

Execution: python longest_consecutive.py
"""
import unittest
from typing import List


def longest_consecutive_bf(nums: List[int]):
    """Function for brute-force longest consecutive."""
    longest_streak = 0

    for num in nums:
        current_num = num
        current_streak = 1

        while current_num + 1 in nums:
            current_num += 1
            current_streak += 1

        longest_streak = max(longest_streak, current_streak)

    return longest_streak


def longest_consecutive_sorting(nums: List[int]):
    """Function for sorting approach for longest consecutive."""
    if not nums:
        return 0

    nums.sort()

    longest_streak = 1
    current_streak = 1

    for i in range(1, len(nums)):
        if nums[i] != nums[i-1]:
            if nums[i] == nums[i-1]+1:
                current_streak += 1
            else:
                longest_streak = max(longest_streak, current_streak)
                current_streak = 1
    return max(longest_streak, current_streak)


def longest_consecutive_hashset(nums: List[int]):
    """Function for hashset approach for longest consecutive."""
    longest_streak = 0
    num_set = set(nums)

    for num in num_set:
        if num - 1 not in num_set:
            current_num = num
            current_streak = 1

            while current_num + 1 in num_set:
                current_num += 1
                current_streak += 1

            longest_streak = max(longest_streak, current_streak)

    return longest_streak


class TestLongestConsecutive(unittest.TestCase):
    """Unit test for longest consecutive."""

    def test_bf(self):
        self.assertEqual(longest_consecutive_bf([100, 4, 200, 1, 3, 2]), 4)
        print("Brute-force approach: The longest consecutive elements sequence is [1, 2, 3, 4]. Therefore its length is 4.")

    def test_sorting(self):
        self.assertEqual(longest_consecutive_sorting([100, 4, 200, 1, 3, 2]), 4)
        print("Sorting approach: The longest consecutive elements sequence is [1, 2, 3, 4]. Therefore its length is 4.")

    def test_hashset(self):
        self.assertEqual(longest_consecutive_hashset([100, 4, 200, 1, 3, 2]), 4)
        print("Hashset approach: The longest consecutive elements sequence is [1, 2, 3, 4]. Therefore its length is 4.")


if __name__ == '__main__':
    unittest.main()
